### PR TITLE
Rework Data Model section.

### PR DESCRIPTION
**Pull request by msporny**
This PR implements a number of resolutions made during the last DID WG virtual face-to-face meeting. Namely, it adds different classes of properties, defines all datatypes, and moves representation authoring rules to later in document.

Reviewers should look at the rewritten Data Model section here:

https://pr-preview.s3.amazonaws.com/w3c/did-core/pull/455.html#data-model

and the representation authoring guidelines that was moved here:

https://pr-preview.s3.amazonaws.com/w3c/did-core/pull/455.html#creating-representations

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/did-core/pull/455.html" title="Last updated on Nov 17, 2020, 4:49 PM UTC (ecda543)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/did-core/455/b426d48...ecda543.html" title="Last updated on Nov 17, 2020, 4:49 PM UTC (ecda543)">Diff</a>

### CODE REVIEW

File: index.html
Comment (dlongley) [inline index.html:1229]: We might want to mention IEEE 754 here?

File: index.html
Comment (dlongley) [inline index.html:1218]: I don't think we need `decimal` here -- we should remove and just use `double` as that's the most interoperable mechanism we have for real number representation (we have `double` defined below in this PR). If we want to have integers we can do that ... where the range would need to be constrained by what JavaScript supports (`MAX_SAFE_INTEGER`) as I believe it's likely the lowest common denominator.

File: index.html
Comment (msporny) [reply index.html]: XMLSCHEMA11-2 refers to IEEE754, I also pointed to the warning about numbers in https://tools.ietf.org/html/rfc7159#section-6.

File: index.html
Comment (msporny) [reply index.html]: Agreed, used integer instead, and pointed to the warning about numbers in https://tools.ietf.org/html/rfc7159#section-6.

**Comment by iherman** [comment]
Would it be possible to merge this branch into #454 (or merge #454 into this one)? This PR misses the rules on representation specific properties (ie, [resolution 5](https://www.w3.org/2019/did-wg/Meetings/Minutes/2020-11-05-did#resolution5) and [resolution 6](https://www.w3.org/2019/did-wg/Meetings/Minutes/2020-11-05-did#resolution6)) while those are the focus of #454. These are the two sides of the same coin, and it is a bit difficult to comment on each of them separately...

**Review by kdenhartog — APPROVED** [review]

File: index.html
Comment (OR13) [inline index.html:2358]: I think we need to provide an "example" of how to do this. https://github.com/w3c/did-spec-registries/issues/154

**Review by OR13 — APPROVED** [review]

File: index.html
Comment (brentzundel) [inline index.html:1143]: I like the two buckets. I'm not sure if we want to bikeshed the names here, or whether we should worry about that after the primary changes in the spec (i.e., after this PR is merged).

File: index.html
Comment (brentzundel) [inline index.html:2337]:
```suggestion
applications might require the creation of a new representation. All
```

File: index.html
Comment (brentzundel) [inline index.html:2342]:
```suggestion
A representation MUST define an unambiguous encoding and decoding for all
```

File: index.html
Comment (brentzundel) [inline index.html:2345]:
```suggestion
This enables anything that can be represented in the <a>DID document</a> data
model to also be represented in a compliant representation.
```

File: index.html
Comment (brentzundel) [reply index.html:2358]: I agree, but think that would be best handled in another PR.

**Review by brentzundel — APPROVED** [review]
Some minor editorial suggestions, otherwise LGTM.

File: index.html
Comment (msporny) [reply index.html:1143]: I expect we'll bike shed the names in another PR. This PR assumes that we'll do that later.

File: index.html
Comment (msporny) [reply index.html:2358]: Agreed. Also note that the "contents" of those specifications have 3 examples in DID Core: JSON, JSON-LD, and CBOR.

**Comment by msporny** [comment]
@iherman wrote:
> Would it be possible to merge this branch into #454 (or merge #454 into this one)?

I was trying to keep the PR fairly small, dealing w/ the base set of changes to the data model in this PR, and then @dlongley adding on top of it to address other resolutions made (that he was asked to write a PR for).

Now that we have multiple approvals for this PR, I'm hesitant to do that @iherman because it could affect the review results. I know it's a pain to review, but I'd rather keep the two separate given that we have multiple positive reviews with changes made for those reviews. We'll have to deal with #454 as a separate PR.

**Comment by kdenhartog** [comment]
> @iherman wrote:
> 
> > Would it be possible to merge this branch into #454 (or merge #454 into this one)?
> 
> I was trying to keep the PR relatively small and focused, dealing w/ the base set of changes to the data model per the virtual face to face in this PR, and then @dlongley adding on top of it to address other resolutions made (that he was asked to write a PR for).
> 
> Now that we have multiple approvals for this PR, I'm hesitant to do that @iherman because it could affect the review results. I know it's a pain to review, but I'd rather keep the two separate given that we have multiple positive reviews with changes made based on those reviews. We'll have to deal with #454 as a separate PR.

It appears all approvals from Orie, Brent, myself, and Dave (author of #454 and approver of this PR) on this PR have also approved the other PR if it makes it easier to get both through with fewer merge conflicts. Hard to say what other reviewers will view as controversial though and so the age old wisdom of keeping PRs small and focused would lean me towards keeping separate as well.

File: index.html
Comment (talltree) [inline index.html:2333]: Suggest changing this heading to "Representation Requirements" but it's okay to do that in a subsequent PR as well.

**Review by talltree — APPROVED** [review]
Major improvement. Let's pull in ASAP. Can be tweaked by subsequent PRs.

**Review by peacekeeper — APPROVED** [review]
For the record, I continue to believe that it's a mistake to include representation-specific syntax in the abstract data model and other representations (e.g. to see `@context` in a YAML file). But I agree this PR reflects the decisions recently made during the TPAC DID WG F2F.

**Review by rhiaro — APPROVED** [review]

File: index.html
Comment (jricher) [inline index.html:2343]: This should be defined in terms of the types associated with the property values and not the values themselves. It's a subtle difference, but it allows us to continue to talk about things in terms of property definitions.

File: index.html
Comment (jricher) [inline index.html:2348]:
```suggestion
The representation MUST be uniquely associated with an IANA-registered MIME type.
```

File: index.html
Comment (jricher) [inline index.html:2355]: The representation must also define default generic type processing rules for consumption of all data types within the representation.

**Review by jricher — CHANGES_REQUESTED** [review]

File: index.html
Comment (brentzundel) [inline index.html:2345]: @jricher would wording this address your concern?
```suggestion
A representation MUST define an unambiguous encoding and decoding for all
property names and the types of their associated values, as defined in this
specification. This enables anything that can be represented in the
<a>DID document</a> data model to also be representable in a compliant
representation.
```

File: index.html
Comment (brentzundel) [reply index.html:2348]: just to clarify, would this mean that if a representation is already associated with `application/did+foo` then no other representation could come along and use that mime type?

File: index.html
Comment (brentzundel) [reply index.html:2355]: I'm not understanding how this differs from a representation defining unambiguous encoding and decoding for all property names and the data types of their associated values, could you clarify?

File: index.html
Comment (msporny) [reply index.html:2343]: Agree. I just moved the text in this PR and expect that we'll make the change you suggest in a subsequent PR, when the "Representation Requirements" section is updated with rewording / new rules that we might have missed (like how you register a representation in the DID Spec registries).

File: index.html
Comment (msporny) [reply index.html:2348]: Yes, @jricher's change is correct and yes, that means that each MIME type is associated with one set of representation rules.

File: index.html
Comment (msporny) [reply index.html:2355]: I think I understand what @jricher is getting at, and that's planned for a future PR. Basically, each representation must list all the abstract data model data types and how you express those concretely in the representation. The representation sections kinda/sorta have this right now, but if this PR goes in, we'll be able to be far more explicit and prescriptive about those serializations in each representation.

File: index.html
Comment (msporny) [inline index.html:2333]:
```suggestion
      <h2>Representation Requirements</h2>
```

Comment (msporny) [reply index.html]: Done.

File: index.html
Comment (msporny) [inline index.html:2345]:
```suggestion
A representation MUST define an unambiguous encoding and decoding for all
data model <a href="#data-model">data types</a> defined in this specification.
This enables anything that can be represented in the <a>DID document</a> data
model to also be represented in a compliant representation.
```

Comment (msporny) [reply index.html:2345]: Perhaps this is what @jricher is looking for?

Comment (jricher) [reply index.html:2345]: More like this, translating the property names is also important (this is what lets JSON-LD have additional contexts and namespaces, and CBOR have tags for property names):

```suggestion
A representation MUST define an unambiguous encoding and decoding for all
property names and all data model <a href="#data-model">data types</a> 
as defined in this specification.
This enables anything that can be represented in the <a>DID document</a> data
model to also be represented in a compliant representation.
```

File: index.html
Comment (jricher) [reply index.html:2355]: @msporny has the right of it, but it also needs to clearly go both ways, to and from the data model data types and the representation's data types. So you need things like this strawman:

"If you see an Integer in the ADM it goes in a JSON Number"

As well as:

"If you see a JSON Number in the representation, and don't have an additional property definition, it is an ADM Float".

I'm ok with this being in a separate PR but this can't be lost.

**Review by brentzundel — APPROVED** [review]

**Comment by msporny** [comment]
Normative, multiple reviews, changes requested and made, no objections, [RESOLUTION to merge](https://www.w3.org/2019/did-wg/Meetings/Minutes/2020-11-17-did#resolution2), merging.

**Comment by iherman** [comment]
The issue was discussed in a [meeting](https://www.w3.org/2019/did-wg/Meetings/Minutes/2020-11-24-did#section3) on 2020-11-24)

- no resolutions were taken

<details><summary><i>View the transcript</i></summary>

### 3. PR 454 and 455... 

_See github pull request [#454](https://github.com/w3c/did-core/pull/454), [#455](https://github.com/w3c/did-core/pull/455)._




**Brent Zundel:** status update?  

**dlongely:** we are waiting for a response from mike jones regarding the suggestions for adding a note to the PR / regardign @context in the JSON representation section.  
… unsure if he is still objecting  

**Brent Zundel:** I will ping him again  

**Markus Sabadello:** after merging 455 and then 454 we still need to revisit the exact terminology... lots of different properties type to hammer down.  

_See github issue [#463](https://github.com/w3c/did-core/issues/463)._




**Markus Sabadello:** feedback welcome  

**Brent Zundel:** other PRs that need to be discussed in this meeting?  

</details>

**Pull request merged**
